### PR TITLE
Issue #5664 Client Accounts.forgotPassword and .verifyEmail pass error in callback

### PR DESCRIPTION
**Pull request by mishavee**
To resolve issue #5664.
Accounts.forgotPassword and Accounts.verifyEmail to return error in callback instead of exception.
bugfix in packages/accounts-password/password_client.js
Two additional tests in packages/accounts-password/password_tests.js

### CODE REVIEW

File: packages/accounts-password/password_client.js [hunk 1]
```diff
@@ -208,8 +208,10 @@ Accounts.changePassword = function (oldPassword, newPassword, callback) {
  * @param {Function} [callback] Optional callback. Called with no arguments on success, or with a single `Error` argument on failure.
  */
 Accounts.forgotPassword = function(options, callback) {
-  if (!options.email)
-    throw new Error("Must pass options.email");
+  if (!options.email) {
+    callback(new Meteor.Error(400, "Must pass options.email"));
```
Comment (zimme) [inline packages/accounts-password/password_client.js]: `callback` is optional, you should check that it's available before using it.
Should we throw if there is no `callback`?

File: packages/accounts-password/password_client.js [hunk 2]
```diff
@@ -255,8 +257,10 @@ Accounts.resetPassword = function(token, newPassword, callback) {
  * @param {Function} [callback] Optional callback. Called with no arguments on success, or with a single `Error` argument on failure.
  */
 Accounts.verifyEmail = function(token, callback) {
-  if (!token)
-    throw new Error("Need to pass token");
+  if (!token) {
+    callback(new Meteor.Error(400, "Need to pass token"));
```
Comment (zimme) [inline packages/accounts-password/password_client.js]: Same as previous comment

**Comment by mishavee** [comment]
Good pickup @zimme, I've altered the code to only use the callback if it is passed in.
This applies to forgotPassword, verifyEmail, createUser, changePassword, resetPassword.
I've updated tests to cover this too.

**Comment by stubailo** [comment]
Hey, this looks great. Unfortunately, I don't have time to look at this in depth right now. I'd suggest one change off the bat, which is that another option would be to throw the type error if a callback is not passed. Otherwise, you'd going to have a function call silently fail with no way for the developer to find out? Or maybe that's the point here, I'm not sure.

**Comment by mishavee** [comment]
Thanks for the feedback @stubailo :)  I think the point is the error is either of interest to the developer, or ignored ... not great practice, but seems appropriate. I'm happy to have another way!

File: packages/accounts-password/password_client.js [hunk 3]
```diff
@@ -116,7 +116,7 @@ Accounts.createUser = function (options, callback) {
   if (typeof options.password !== 'string')
     throw new Error("options.password must be a string");
```
Comment (zimme) [inline packages/accounts-password/password_client.js]: Shouldn't this be handled by the callback too?

Comment (mishavee) [inline packages/accounts-password/password_client.js]: I think type checking errors need to be thrown as exceptions to get noticed during development, so the developer will realise they're sending in the wrong thing when they test it.
On the other hand, unhandled exceptions in production are ugly for the users.

**Comment by tmeasday** [comment]
Thanks for this PR, sorry we never got around to merging it. We are currently cleaning up the repository and removing old PRs; if you think this PR still has value and are interested in resurrecting it, we encourage you to re-submit it against current `devel`.

**Pull request closed without merging**